### PR TITLE
Fixes #609 Publication with columns

### LINKED ISSUE #609: Publication creation doesn't support specifying columns for published table
##### SUMMARY

Publication creation doesn't support specifying columns for which publication should be created - 
example SQL - `CREATE PUBLICATION users_filtered FOR TABLE users (user_id, firstname)` 

Right now if I pass as a table name  `users (user_id, firstname)` it will quote the whole thing, which ends up as invalid SQL
 
##### ISSUE TYPE
- Missing functionality

##### COMPONENT NAME

community.postgresql.postgresql_publication

##### ADDITIONAL INFORMATION

It should allow to set-up publication with a limited column set - since it is possible in Postgres >= 15

Specs could look like this:
```yaml
tables:
  - "public"."users" (user_id, firstname)
```
or
```yaml
tables:
  - users:
    - user_id
    - firstname
```

### PR DESCRIPTION
**Pull request by kacperk**
##### SUMMARY

Fixes #609

Add possibility of creating publication with only specific columns

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

postgresql_publication

##### ADDITIONAL INFORMATION

```paste below
    postgresql_publication:
      name: '{{ test_pub }}'
      columns:
        users:
          - user_id
          - email
        some_other_table_with_al_columns: []
```

### CODE REVIEW

**Comment by kacperk** [comment]
I can't start integration test locally - so it is not finished. I'll need some help with that.

I'm getting `ERROR! couldn't resolve module/action 'locale_gen'.` when trying to run it.

File: plugins/modules/postgresql_publication.py [hunk 1]
```diff
@@ -326,13 +363,15 @@ def check_pub(self):
             # FOR TABLES IN SCHEMA statement is supported since PostgreSQL 15
             if self.pg_srv_ver >= 150000:
                 self.attrs['schemas'] = self.__get_schema_pub_info()
+            if self.pg_srv_ver >= 150000:
+                self.attrs['columns'] = self.__get_columns_pub_info()
```
Comment (hunleyd) [inline plugins/modules/postgresql_publication.py:367]: update the comment and lose the duplicate if?
```suggestion
            # FOR TABLES IN SCHEMA statement is supported since PostgreSQL 15, as is publishing by column
            if self.pg_srv_ver >= 150000:
                self.attrs['schemas'] = self.__get_schema_pub_info()
                self.attrs['columns'] = self.__get_columns_pub_info()
```

**Review by hunleyd — CHANGES_REQUESTED** [review]
please add a changelog fragment

File: plugins/modules/postgresql_publication.py [hunk 2]
```diff
@@ -775,6 +875,8 @@ def main():
 
     if tables_in_schema is not None and pg_srv_ver < 150000:
         module.fail_json(msg="Publication of tables in schema is supported by PostgreSQL 15 or greater")
+    if columns is not None and pg_srv_ver < 150000:
+        module.fail_json(msg="Publication of columns is supported by PostgreSQL 15 or greater")
```
Comment (Andersson007) [inline plugins/modules/postgresql_publication.py:879]:
```suggestion
    if columns and pg_srv_ver < 150000:
        module.fail_json(msg="Publication of columns is supported by PostgreSQL 15 or greater")
```
Does it makes sense to pass an empty dict to the argument or we can simplify this a bit like suggested?

File: tests/integration/targets/postgresql_publication/tasks/postgresql_publication_initial.yml [hunk 3]
```diff
@@ -496,6 +497,27 @@
...
+    <<: *task_parameters
+    postgresql_publication:
+      <<: *pg_parameters
+      name: '{{ test_pub }}'
+      columns:
+        '{{ test_table1 }}': []
+        '{{ test_schema }}.{{ test_table2 }}': ['name']
+        '{{ test_table3 }}': ['id']
+      trust_input: false
+    check_mode: true
```
Comment (Andersson007) [inline tests/integration/targets/postgresql_publication/tasks/postgresql_publication_initial.yml:513]:
```suggestion
    check_mode: true
```
1. Could you also add same task but without check_mode (i.e. in real mode) and check the result has actually changed?
2. Could you then add **the same task** in real mode and then check if nothing has changed
3. Then add another task changing the columns a little bit, like:
  - add another column to `'{{ test_schema }}.{{ test_table2 }}': ['name']`
  - pass an empty list to  `'{{ test_table3 }}': ['id']` and see what'll happen
 
4. Add other changes in other tasks
5. To check the actual state you, in addition to what the module return, can use `postgresql_query` module.

We need to cover majority of use cases with the tests as the changes are quite significant.

**Review by Andersson007 — COMMENTED** [review]
@kacperk hello, thanks for the PR

**Comment by Andersson007** [comment]
> I can't start integration test locally - so it is not finished. I'll need some help with that.
> 
> I'm getting `ERROR! couldn't resolve module/action 'locale_gen'.` when trying to run it.

just mark this PR as draft here and push: you'll get the result of the tests here

**Comment by kacperk** [comment]
@Andersson007 I've manage to run the tests and I've added more of them. I think the PR is ready now.

File: tests/integration/targets/postgresql_publication/tasks/postgresql_publication_initial.yml [hunk 4]
```diff
@@ -496,11 +497,324 @@
...
+        - result.queries == ["CREATE PUBLICATION \"test_pub_with_columns\" FOR TABLE \"public\".\"{{ test_table1 }}\", \"{{ test_schema }}\".\"{{ test_table2 }}\" (\"name\")"]
+    when: postgres_version_resp.stdout is version('15', '>=')
+
+    # Check
+  - name: postgresql_publication - check that nothing has been changed
+    when: postgres_version_resp.stdout is version('15', '>=')
+    <<: *task_parameters
+    postgresql_query:
+      <<: *pg_parameters
+      query: SELECT * FROM pg_publication WHERE pubname = 'test_pub_with_columns'
```
Comment (Andersson007) [inline tests/integration/targets/postgresql_publication/tasks/postgresql_publication_initial.yml:559]:
```suggestion
      query: SELECT * FROM pg_publication WHERE pubname = 'test_pub_with_columns'
    when: postgres_version_resp.stdout is version('15', '>=')
```

**Review by Andersson007 — COMMENTED** [review]
@kacperk if you press `Commit suggestion` in all my suggestions, it'll probably fix the failed tests

File: plugins/modules/postgresql_publication.py [hunk 5]
```diff
@@ -210,19 +226,11 @@
...
-    exec_sql,
-    get_conn_params,
-    get_server_version,
-    pg_cursor_args,
-    postgres_common_argument_spec,
-    set_comment,
-)
+    connect_to_db, ensure_required_libs, exec_sql, get_conn_params,
+    get_server_version, pg_cursor_args, postgres_common_argument_spec,
+    set_comment)
```
Comment (Andersson007) [inline plugins/modules/postgresql_publication.py:233]:
```suggestion
    connect_to_db, ensure_required_libs, exec_sql, get_conn_params,
    get_server_version, pg_cursor_args, postgres_common_argument_spec,
    set_comment)
```
does it anyhow relate to the PR? could you please put it back. Even in places where I agree with refactoring, it should go to a separate PR. Let's not mix unrelated things in a single PR

Comment (kacperk) [reply plugins/modules/postgresql_publication.py:233]: @Andersson007 I followed the contributing guide and this is a change applied by the pre-commit hook.

Comment (Andersson007) [reply plugins/modules/postgresql_publication.py:233]: oh, sorry, we should remove that thing from the guide - it does too much.. thanks for following the guide

**Review by hunleyd — APPROVED** [review]

**Review by Andersson007 — APPROVED** [review]

**Comment by Andersson007** [comment]
@kacperk thanks for the contribution! Great module expansion!
@hunleyd thanks for reviewing!

@kacperk 
- feel free to add yourself to the `authors:` field in the module DOCUMENTATION section in another PR
- hope you'll be around if someone submits related issues
- we have a forum group and a channel on Matrix, you're welcome to join, see the [communication](https://github.com/ansible-collections/community.postgresql#communication) guide for details. There's not a lot of messages coming from them, mostly asking for help with releasing and announcements
